### PR TITLE
Bulk delete properties

### PR DESCRIPTION
**Pull request by diegogrz**
Added methods to support deleting multiple properties on a single ES request to minimize collisions.

Fixes: https://github.com/v5analytics/visallo-lts/issues/1983

### CODE REVIEW

File: accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java [hunk 1]
```diff
@@ -361,26 +362,19 @@ void saveProperties(
...
+            propertySoftDeletes.forEach( p -> { propertyList.add(new PropertyDescriptor(p.getKey(), p.getName(), p.getVisibility()));});
+
+            getSearchIndex().deleteProperties(
+                    this,
+                    element,
+                    propertyList,
+                    authorizations
+            );
+
+            // Add new element
```
Comment (joeferner) [inline accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java]: remove unneeded comment and extra white space above.

File: accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java [hunk 2]
```diff
@@ -361,26 +362,19 @@ void saveProperties(
...
-                        element,
-                        propertySoftDeleteMutation.getKey(),
-                        propertySoftDeleteMutation.getName(),
-                        propertySoftDeleteMutation.getVisibility(),
-                        authorizations
-                );
-            }
+            // Bulk delete properties
+            List<PropertyDescriptor> propertyList = Lists.newArrayList();
+            propertyDeletes.forEach( p -> { propertyList.add(new PropertyDescriptor(p.getKey(), p.getName(), p.getVisibility()));});
```
Comment (joeferner) [inline accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java]: use `PropertyDescriptor.from`

File: accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java [hunk 3]
```diff
@@ -402,7 +396,11 @@ void deleteProperty(AccumuloElement element, Property property, Authorizations a
         elementMutationBuilder.addPropertyDeleteToMutation(m, property);
         addMutations(getWriterFromElementType(element), m);
 
-        getSearchIndex().deleteProperty(this, element, property, authorizations);
+        getSearchIndex().deleteProperty(
+                this,
+                element,
+                new PropertyDescriptor(property.getKey(), property.getName(), property.getVisibility()),
```
Comment (joeferner) [inline accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java]: use `PropertyDescriptor.from`

File: accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java [hunk 4]
```diff
@@ -414,7 +412,11 @@ void softDeleteProperty(AccumuloElement element, Property property, Authorizatio
         elementMutationBuilder.addPropertySoftDeleteToMutation(m, property);
         addMutations(getWriterFromElementType(element), m);
 
-        getSearchIndex().deleteProperty(this, element, property, authorizations);
+        getSearchIndex().deleteProperty(
+                this,
+                element,
+                new PropertyDescriptor(property.getKey(), property.getName(), property.getVisibility()),
```
Comment (joeferner) [inline accumulo/src/main/java/org/vertexium/accumulo/AccumuloGraph.java]: use `PropertyDescriptor.from`

File: core/src/main/java/org/vertexium/search/PropertyDescriptor.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import org.vertexium.Property;
+import org.vertexium.Visibility;
+
+/**
+ * Encapsulates the parameters necessary to perform a Property search
+ */
+public class PropertyDescriptor {
+    private String key;
+    private String name;
+    private Visibility visibility;
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/search/PropertyDescriptor.java]: Can you make these fields `final` and remove the setters

File: core/src/main/java/org/vertexium/search/PropertyDescriptor.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
+package org.vertexium.search;
+
+import org.vertexium.Property;
+import org.vertexium.Visibility;
+
+/**
+ * Encapsulates the parameters necessary to perform a Property search
+ */
+public class PropertyDescriptor {
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/search/PropertyDescriptor.java]: Can you add `toString` and `equals` methods to this class.

File: core/src/main/java/org/vertexium/search/PropertyDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
+package org.vertexium.search;
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/search/PropertyDescriptor.java]: Change package to `org.vertexium`

File: core/src/main/java/org/vertexium/search/SearchIndex.java [hunk 8]
```diff
@@ -6,19 +6,33 @@
...
 public interface SearchIndex {
     void addElement(Graph graph, Element element, Authorizations authorizations);
 
     void deleteElement(Graph graph, Element element, Authorizations authorizations);
 
-    void deleteProperty(Graph graph, Element element, Property property, Authorizations authorizations);
+    /**
+     * Default delete property simply calls deleteProperty in a loop. It is up to the SearchIndex implementation to decide
+     * if a collective method can be made more efficient
+     * @param graph
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/search/SearchIndex.java]: remove `@param`s

File: elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java [hunk 9]
```diff
@@ -390,19 +393,11 @@ public void alterElementVisibility(
...
-                        "ctx._source.remove(oldFieldName);",
-                        ScriptService.ScriptType.INLINE
-                )
-                .setScriptParams(ImmutableMap.of(
-                        "oldFieldName", oldFieldName
-                ))
-                .get();
+        esRemoveFields(element, oldFieldName);
+
+        // Add new element
```
Comment (joeferner) [inline elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java]: remove redundent comment

File: elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java [hunk 10]
```diff
@@ -390,19 +393,11 @@ public void alterElementVisibility(
...
-                .setType(ELEMENT_TYPE)
-                .setScript(
-                        "ctx._source.remove(oldFieldName);",
-                        ScriptService.ScriptType.INLINE
-                )
-                .setScriptParams(ImmutableMap.of(
-                        "oldFieldName", oldFieldName
-                ))
-                .get();
+        esRemoveFields(element, oldFieldName);
```
Comment (joeferner) [inline elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java]: rename to `removeFields` or `removeFieldsFromDocument`

File: elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java [hunk 11]
```diff
@@ -1107,6 +1101,34 @@ public void flush(Graph graph) {
...
+     * Helper method to remove fields from source. This method will generate a ES update request. Retries on conflict.
+     *
+     * @param element Element that can be mapped to an ES document
+     * @param fields fields to remove
+     */
+    private void esRemoveFields(Element element, Collection<String> fields) {
+        String script = "";
+        String separator = "";
+
+        for (String field : fields) {
```
Comment (joeferner) [inline elasticsearch-singledocument/src/main/java/org/vertexium/elasticsearch/ElasticsearchSingleDocumentSearchIndex.java]: can be rewritten `fields.stream().map(...).collect(Collectors.joining("; "));`

File: inmemory/src/main/java/org/vertexium/inmemory/InMemoryElement.java [hunk 12]
```diff
@@ -480,10 +482,16 @@ protected void saveMutationToSearchIndex(
...
     ) {
         if (alterPropertyVisibilities != null && alterPropertyVisibilities.size() > 0) {
-            for (AlterPropertyVisibility apv : alterPropertyVisibilities) {
-                Visibility existingVisibility = apv.getExistingVisibility();
-                getGraph().getSearchIndex().deleteProperty(getGraph(), element, apv.getKey(), apv.getName(), existingVisibility, authorizations);
-            }
+
+            // Bulk delete
+            List <PropertyDescriptor> propertyList = Lists.newArrayList();
+            alterPropertyVisibilities.forEach( p -> { propertyList.add(new PropertyDescriptor(p.getKey(), p.getName(), p.getExistingVisibility()));});
```
Comment (joeferner) [inline inmemory/src/main/java/org/vertexium/inmemory/InMemoryElement.java]: use `PropertyDescriptor.from`

File: core/src/main/java/org/vertexium/search/PropertyDescriptor.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+        this.key = key;
+        this.name = name;
+        this.visibility = visibility;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    public static PropertyDescriptor from(Property p) {
```
Comment (joeferner) [inline core/src/main/java/org/vertexium/search/PropertyDescriptor.java]: rename to `fromProperty`

**Comment by diegogrz** [comment]
The testChangeVisibilityVertex test changes the visibility on a Vertex which causes the index deleteProperties method to get executed

**Pull request merged**